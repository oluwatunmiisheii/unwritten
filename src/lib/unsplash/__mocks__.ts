import type { UploadedBy, Image, ImageResult } from "./unsplash.model"
import { v4 as uuidv4 } from 'uuid';

export const createMockUploadedBy = (overrides: Partial<UploadedBy> = {}): UploadedBy => ({
    first_name: 'John',
    last_name: 'Doe',
    id: 'user-123',
    username: 'johndoe',
    profile_image: {
        large: 'https://example.com/large-profile.jpg',
        medium: 'https://example.com/medium-profile.jpg',
        small: 'https://example.com/small-profile.jpg',
        ...overrides.profile_image
    },
    ...overrides
})

export const createMockImage = (overrides: Partial<Image> = {}): Image => ({
    id: uuidv4(),
    slug: 'example-image',
    width: 1920,
    height: 1080,
    color: '#000000',
    description: 'A sample image',
    alt_description: 'Example alt text',
    likes: 100,
    urls: {
        full: 'https://example.com/full-image.jpg',
        raw: 'https://example.com/raw-image.jpg',
        regular: 'https://example.com/regular-image.jpg',
        small: 'https://example.com/small-image.jpg',
        small_s3: 'https://example.com/small-s3-image.jpg',
        thumb: 'https://example.com/thumb-image.jpg',
        ...overrides.urls
    },
    user: createMockUploadedBy(overrides.user),
    ...overrides
})

export const createMockImageResult = (
    imageCount = 3,
    overrides: Partial<ImageResult> = {}
): ImageResult => ({
    total: imageCount,
    total_pages: Math.ceil(imageCount / 10),
    results: Array.from({ length: imageCount }, () => createMockImage({})),
    ...overrides
})