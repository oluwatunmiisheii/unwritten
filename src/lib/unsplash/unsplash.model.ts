export interface UploadedBy {
    first_name: string;
    id: string;
    last_name: string;
    profile_image: {
        large: string;
        medium: string;
        small: string;
    };
    username: string;
}

export interface Image {
    id: string;
    slug: string;
    width: number;
    height: number;
    color: string;
    description?: string;
    likes: number;
    alt_description?: string;
    urls: {
        full: string;
        raw: string;
        regular: string;
        small: string;
        small_s3: string;
        thumb: string;
    };
    user: UploadedBy;
}

export interface ImageResult {
    total: number;
    total_pages: number;
    results: Image[];
}