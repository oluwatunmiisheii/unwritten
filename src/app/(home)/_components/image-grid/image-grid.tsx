'use client'

import { useState } from 'react'
import { ImageResult } from '@/lib/unsplash/unsplash.model'
import Image from 'next/image'
import React from 'react'
import { getImages } from '@/lib/unsplash/unsplash.service'

interface ImageGridProps {
  initialImages: ImageResult['results']
  searchQuery: string
}

export const ImageGrid = ({ initialImages, searchQuery }: ImageGridProps) => {
  const [images, setImages] = useState(initialImages)
  const [page, setPage] = useState(1)
  const [hasMore, setHasMore] = useState(true)
  const [isLoading, setIsLoading] = useState(false)

  const loadMore = async () => {
    if (isLoading) return

    setIsLoading(true)
    const nextPage = page + 1

    try {
      const result = await getImages(nextPage, searchQuery)

      if (result.results.length > 0) {
        setImages((prev) => [...prev, ...result.results])
        setPage(nextPage)
        setHasMore(nextPage < result.total_pages)
      } else {
        setHasMore(false)
      }
    } catch (error) {
      console.warn(error)
    } finally {
      setIsLoading(false)
    }
  }

  return (
    <div className="mt-20">
      <h5 className="text-2xl mb-8">
        Showing results for{' '}
        <span className="font-bold border-b border-dotted border-black capitalize">
          {searchQuery}
        </span>
      </h5>
      <div className="grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-4 mt-4">
        {images.map((image, index) => (
          <div
            className="flex w-full flex-col items-center border border-gray-200 overflow-hidden rounded-lg bg-zinc-100 shadow-sm"
            key={image.id}
          >
            <div className="w-full relative h-52">
              <Image
                src={image.urls.regular}
                priority={index < 10}
                alt={image.description || image.alt_description || image.slug}
                fill
                sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 25vw"
                className="object-cover object-top"
              />
              <div className="bg-white shadow rounded-full absolute text-[11px] px-2.5 py-1 right-3 top-3">
                ❤️ {image.likes}
              </div>
            </div>
            <div className="p-5 flex items-between w-full items-center justify-between">
              <div className="size-8 rounded-full border-2 border-zinc-800 flex-shrink-0">
                <Image
                  src={image.user.profile_image.large}
                  alt={image.user.username}
                  height={32}
                  width={32}
                  className="rounded-full"
                />
              </div>
              <p className="text-sm truncate pl-2">{image.user.username}</p>
            </div>
          </div>
        ))}
      </div>
      {hasMore && (
        <div className="flex justify-center mt-8">
          <button
            type="button"
            className="rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
            onClick={loadMore}
          >
            Load More
          </button>
        </div>
      )}
    </div>
  )
}
