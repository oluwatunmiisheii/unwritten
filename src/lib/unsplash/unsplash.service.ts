"use server";

import { ImageResult } from "@/lib/unsplash/unsplash.model";
import { get } from "@/lib/fetch";


export async function getImages(page = 1, query = "Manchester"): Promise<ImageResult> {
    try {
        return get<ImageResult>(`/search/photos?query=${query}&page=${page}&per_page=16`, {
            options: {
                headers: {
                    Authorization: `Client-ID ${process.env.UNSPLASH_ACCESS_KEY}`,
                },
            },
        });
    } catch (error) {
        console.log(error);
        throw error;
    }
}