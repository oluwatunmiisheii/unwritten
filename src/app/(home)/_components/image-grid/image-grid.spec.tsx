import React, { ComponentProps } from 'react'
import { render, screen } from '@testing-library/react'
import { describe, test, expect, vi, beforeEach } from 'vitest'
import { getImages } from '@/lib/unsplash/unsplash.service'
import { ImageGrid } from './image-grid'
import { createMockImageResult } from '@/lib/unsplash/__mocks__'
import userEvent from '@testing-library/user-event'

vi.mock('@/lib/unsplash/unsplash.service', () => ({
  getImages: vi.fn(),
}))

describe('ImageGrid', () => {
  const defaultProps = {
    initialImages: createMockImageResult().results,
    searchQuery: 'test',
    initialPage: 1,
  } as ComponentProps<typeof ImageGrid>

  beforeEach(() => {
    vi.clearAllMocks()
  })

  test('renders initial images and search query', () => {
    const { container } = render(<ImageGrid {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('loads more images when "Load More" button is clicked', async () => {
    const user = userEvent.setup()
    const mockNewImages = createMockImageResult().results

    vi.mocked(getImages).mockResolvedValueOnce({
      total: 6,
      results: mockNewImages,
      total_pages: 2,
    })

    render(<ImageGrid {...defaultProps} />)

    const loadMoreButton = screen.getByText('Load More')
    await user.click(loadMoreButton)

    expect(getImages).toHaveBeenCalledWith(2, 'test')
  })

  test('hides "Load More" button when there are no more images', async () => {
    const user = userEvent.setup()

    vi.mocked(getImages).mockResolvedValueOnce({
      total: 6,
      results: createMockImageResult().results,
      total_pages: 2,
    })

    render(<ImageGrid {...defaultProps} />)

    const loadMoreButton = screen.getByText('Load More')
    await user.click(loadMoreButton)

    expect(getImages).toHaveBeenCalledWith(2, 'test')
    expect(loadMoreButton).not.toBeInTheDocument()
  })
})
