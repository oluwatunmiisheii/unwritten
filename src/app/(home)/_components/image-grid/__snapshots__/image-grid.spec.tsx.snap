// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageGrid > renders initial images and search query 1`] = `
<div>
  <div
    class="mt-20"
  >
    <h5
      class="text-2xl mb-8"
    >
      Showing results for
       
      <span
        class="font-bold border-b border-dotted border-black capitalize"
      >
        test
      </span>
    </h5>
    <div
      class="grid grid-cols-4 gap-4 mt-4"
    >
      <div
        class="flex w-full flex-col items-center border border-gray-200 overflow-hidden rounded-lg bg-zinc-100 shadow-sm"
      >
        <div
          class="w-full relative h-52"
        >
          <img
            alt="A sample image"
            class="object-cover object-top"
            data-nimg="fill"
            decoding="async"
            sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 25vw"
            src="/_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=3840&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
          <div
            class="bg-white shadow rounded-full absolute text-[11px] px-2.5 py-1 right-3 top-3"
          >
            ❤️ 
            100
          </div>
        </div>
        <div
          class="p-5 flex items-between w-full items-center justify-between"
        >
          <div
            class="size-8 rounded-full border-2 border-zinc-800 flex-shrink-0"
          >
            <img
              alt="johndoe"
              class="rounded-full"
              data-nimg="1"
              decoding="async"
              height="32"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=64&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=64&q=75 2x"
              style="color: transparent;"
              width="32"
            />
          </div>
          <p
            class="text-sm truncate pl-2"
          >
            johndoe
          </p>
        </div>
      </div>
      <div
        class="flex w-full flex-col items-center border border-gray-200 overflow-hidden rounded-lg bg-zinc-100 shadow-sm"
      >
        <div
          class="w-full relative h-52"
        >
          <img
            alt="A sample image"
            class="object-cover object-top"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 25vw"
            src="/_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=3840&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
          <div
            class="bg-white shadow rounded-full absolute text-[11px] px-2.5 py-1 right-3 top-3"
          >
            ❤️ 
            100
          </div>
        </div>
        <div
          class="p-5 flex items-between w-full items-center justify-between"
        >
          <div
            class="size-8 rounded-full border-2 border-zinc-800 flex-shrink-0"
          >
            <img
              alt="johndoe"
              class="rounded-full"
              data-nimg="1"
              decoding="async"
              height="32"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=64&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=64&q=75 2x"
              style="color: transparent;"
              width="32"
            />
          </div>
          <p
            class="text-sm truncate pl-2"
          >
            johndoe
          </p>
        </div>
      </div>
      <div
        class="flex w-full flex-col items-center border border-gray-200 overflow-hidden rounded-lg bg-zinc-100 shadow-sm"
      >
        <div
          class="w-full relative h-52"
        >
          <img
            alt="A sample image"
            class="object-cover object-top"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 25vw"
            src="/_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=3840&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fexample.com%2Fregular-image.jpg&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
          <div
            class="bg-white shadow rounded-full absolute text-[11px] px-2.5 py-1 right-3 top-3"
          >
            ❤️ 
            100
          </div>
        </div>
        <div
          class="p-5 flex items-between w-full items-center justify-between"
        >
          <div
            class="size-8 rounded-full border-2 border-zinc-800 flex-shrink-0"
          >
            <img
              alt="johndoe"
              class="rounded-full"
              data-nimg="1"
              decoding="async"
              height="32"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=64&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Flarge-profile.jpg&w=64&q=75 2x"
              style="color: transparent;"
              width="32"
            />
          </div>
          <p
            class="text-sm truncate pl-2"
          >
            johndoe
          </p>
        </div>
      </div>
    </div>
    <div
      class="flex justify-center mt-8"
    >
      <button
        class="rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        type="button"
      >
        Load More
      </button>
    </div>
  </div>
</div>
`;
